ICU-21156 Fixed DateIntervalFormatter to work correctly with K and k…

richgillam opened the pull request:
… in the skeleton.

<!--
Thank you for your pull request.
Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html#License
-->

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21156
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added

pedberg-icu commented on icu4c/source/i18n/dtitvinf.cpp:
```diff
@@ -594,20 +594,18 @@ DateIntervalInfo::getBestSkeleton(const UnicodeString& skeleton,
-    if ( skeleton.indexOf(CHAR_Z) != -1 ) {
-        copySkeleton = skeleton;
-        copySkeleton.findAndReplace(UnicodeString(CHAR_Z), UnicodeString(CHAR_V));
+    if ( skeleton.indexOf(LOW_Z) != -1 || skeleton.indexOf(LOW_K) != -1 || skeleton.indexOf(CAP_K) != -1 ) {        copySkeleton = skeleton;
```
missing a CR after the '{'?

richgillam replied:
Whoops.  Fixed!

richgillam commented:
> Please rewrite the commit message to remove the colon in "ICU-21156: Fixed...". The commit message checker is very strict about the format it allows, and colon is not allowed here.

Sorry about that.  Fixed!

richgillam commented on icu4c/source/i18n/dtitvfmt.cpp:
```diff
@@ -1023,11 +1021,8 @@ DateIntervalFormat::getDateTimeSkeleton(const UnicodeString& skeleton,
-    else if ( hCount != 0 ) {
-        normalizedTimeSkeleton.append(LOW_H);
+    if (hourChar != u'\0' ) {
+        normalizedTimeSkeleton.append(hourChar);
```
I think we actually _couldn't_ do it in `normalizedTimeSkeleton`, because I think if we changed it there, we'd lose the fact that the original skeleton had a `K` in it instead of an `h`.  I may be mixed up, but I think the original version of the skeleton is how we do this, and that the only thing normalization does is mess with the field lengths.  But I'm coming to this code fairly cold, so I might not know all the history.  As you point out, though, this is how we were already handling `v` and `z`, so I just piggybacked on top of that.

pedberg-icu replied:
I think you are right.

pedberg-icu commented:
OK now the second commit fails commit message validation because it lacks the ticket. But if you squash (before merge) then you can make that second commit message go away and the commit checks should then pass.

richgillam commented:
>OK now the second commit fails commit message validation because it lacks the ticket. But if you squash (before merge) then you can make that second commit message go away and the commit checks should then pass.

I'm confused.  So if I make changes in response to code review, I have to add the ticket number to each commit message, or always do `git commit --amend`?  And are you saying I need to put up a *new* commit now that squashes these two commits back down into one, or just amend the second commit to include the ticket number?  Or wait until merge time and do a squash merge?

pedberg-icu commented:
> I'm confused. So if I make changes in response to code review, I have to add the ticket number to each commit message, or always do `git commit --amend`? And are you saying I need to put up a _new_ commit now that squashes these two commits back down into one, or just amend the second commit to include the ticket number? Or wait until merge time and do a squash merge?

Well, you need to do one or the other. Either:
1. Every commit message for a ticket needs to have the ticket number; then you can squash at merge time (editing the combined commit message to delete useless info like about fixing the newline). In this case you can go back and amend the commit message for the second commit.
OR
2. You can squash into a single commit before merge and final review, then the single commit that results from the squash should have the ticket number.  I think this is preferred and is usually what I do. There are insturctions for this at http://site.icu-project.org/repository/gitdev; scoll down a couple of pages to "**Options on how to squash:**"; I usually use Option 2. To do this for your case (squashing two commits into one)
- % git rebase -i HEAD~2
- this will bring up a window (probably in vi unless you changed your default command-line editor) with two lines at the top that begin with "pick". Edit the second one to say "squash" instead of "pick", and save the changes.
- this will then bring up another window (also in vi) for you to edit the commit message for the squahsed commit. In this case you can just delete the second commit message (for the newline), keeping just the first (with the ticket number), then save the changes.
- then you need to force push: % git push origin [branchName] --force

richgillam commented:
>You can squash into a single commit before merge and final review, then the single commit that results from the squash should have the ticket number.

Okay, I went back and squashed, and I also read the relevant parts of the page you linked to.  It sounds like in the future, I should address review comments with new commits that contain the ticket number and then, once somebody has given tentative approval, squash the commits together before merging.  Does that sounds right to you?

pedberg-icu commented:
> It sounds like in the future, I should address review comments with new commits that contain the ticket number and then, once somebody has given tentative approval, squash the commits together before merging. Does that sounds right to you?

Yes, that sounds good, and will avoid unnecessary commit check errors.

richgillam commented:
This issue turned out to be the tip of a very big iceberg, which I've addressed in my fix for ICU-21155 (see https://github.com/unicode-org/icu/pull/1293).  I've incorporated this fix into that one, so I'm closing this PR in favor of that one.
